fish-rust/src/common.rs

```diff
@@ -1752,72 +1742,34 @@ pub type Cleanup<T, F> = ScopeGuard<T, F>;
 ///
 /// // hello will be written first, then goodbye.
 /// ```
-pub struct ScopeGuard<T, F: FnOnce(&mut T)> {
-    captured: ManuallyDrop<T>,
-    on_drop: Option<F>,
-}
+pub struct ScopeGuard<T, F: FnOnce(T)>(Option<(T, F)>);
 
-impl<T, F> ScopeGuard<T, F>
-where
-    F: FnOnce(&mut T),
-{
+impl<T, F: FnOnce(T)> ScopeGuard<T, F> {
     /// Creates a new `ScopeGuard` wrapping `value`. The `on_drop` callback is executed when the
     /// ScopeGuard's lifetime expires or when it is manually dropped.
     pub fn new(value: T, on_drop: F) -> Self {
-        Self {
-            captured: ManuallyDrop::new(value),
-            on_drop: Some(on_drop),
-        }
-    }
-
-    /// Cancel the unwind operation, e.g. do not call the previously passed-in `on_drop` callback
-    /// when the current scope expires.
-    pub fn cancel(guard: &mut Self) {
-        guard.on_drop.take();
-    }
-
-    /// Cancels the unwind operation like [`ScopeGuard::cancel()`] but also returns the captured
-    /// value (consuming the `ScopeGuard` in the process).
-    pub fn rollback(mut guard: Self) -> T {
-        guard.on_drop.take();
-        // Safety: we're about to forget the guard altogether
-        let value = unsafe { ManuallyDrop::take(&mut guard.captured) };
-        std::mem::forget(guard);
-        value
-    }
-
-    /// Commits the unwind operation (i.e. applies the provided callback) and returns the captured
-    /// value (consuming the `ScopeGuard` in the process).
-    pub fn commit(mut guard: Self) -> T {
```

Comment:
I'm already using `commit()` because it's the equivalent of the C++ `scoped_push::restore`. We should unify the names